Improve skin downloading: load from `downloadedskins` if possible

Robyt3 opened the pull request:
Previously, skins not found in the `skins` folder were downloaded from the configured URL and saved to the `downloadedskins` folder, but the saved files were never used, leading to wasted downloads and write operations.

Now, if a skin to be downloaded already exists in the `downloadedskins` folder, the modified time of the skin file is determined and included in the HTTP GET request as `If-Modified-Since` header. If the file was not modified on the server since that time, the server will answer with status 304 Not Modified and an empty response body, in which case the existing skin file is loaded. If the skin file was modified, then it will be downloaded like usual. The download will also be retried without including the `If-Modified-Since` header, if the local file is more recent but it failed to be loaded. If a skin could not be downloaded when it should but a file already exists in the `downloadedskins` folder then that file will be loaded instead.

Instead of checking whether the number of skins is different to determine whether to reload the list of skins in the settings menus, the last time that the list of skins was refreshed is now saved and compared, which is more robust. The additional check to refresh the skin list while skins are being downloaded is unnecessary, as any updates to the skin list are now covered by checking the last refresh time.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [X] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

Robyt3 commented on src/game/client/components/skins.cpp:
```diff
@@ -463,15 +425,175 @@ void CSkins::RandomizeSkin(int Dummy)
+	char aPathReal[IO_MAX_PATH_LENGTH];
+	str_format(aPathReal, sizeof(aPathReal), "downloadedskins/%s.png", m_aName);
+
+	const CTimeout Timeout{10000, 0, 8192, 10};
```
I don't really know how to pick the timeout values. The downloaded skins I had were around 16 KiB on average.

Robyt3 commented on src/game/client/components/skins.cpp:
```diff
@@ -463,15 +425,175 @@ void CSkins::RandomizeSkin(int Dummy)
+
+	std::shared_ptr<CHttpRequest> pGet = HttpGet(aUrl);
+	pGet->Timeout(Timeout);
+	pGet->MaxResponseSize(10 * 1024 * 1024); // 10 MiB
```
I think 10 MiB should be a reasonable limit for downloaded skins.

heinrich5991 commented:
> Now, if a skin to be downloaded already exists in the `downloadedskins` folder, the modified time of the skin file is determined and a HTTP HEAD request is sent to the skin download URL to determine when the skin to be downloaded was last modified. The times are compared with each other and if the local file is already up-to-date it will be loaded directly. The skin will be downloaded with a HTTP GET request if the response to the HEAD request does not contain a modified time, if the local file is older, or if the local file is more recent but failed to be loaded. If a skin could not be downloaded when it should but a file already exists in the `downloadedskins` folder then that file will be loaded instead.

Just from reading the description: I think the more correct way to go about this is to use the [`If-Modified-Since`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since) (or even better, but more complicated, the [`If-None-Match`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header). This allows you to do the whole thing in just one request.

Robyt3 commented:
> Just from reading the description: I think the more correct way to go about this is to use the [`If-Modified-Since`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since) (or even better, but more complicated, the [`If-None-Match`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match)) header. This allows you to do the whole thing in just one request.

I changed it to use the `If-Modified-Since` header. Seems like using the `If-None-Match` header would require some separate etag storage and it would also invalidate all existing `downloadedskins` folders.

> (It'd probably be even better to load the old skin unconditionally while the request is in flight and reload it if it changed.)

I changed it to load the PNG from the existing skin file while the HTTP request is running. However, this only saves the time to load the PNG file if skin was not modified. Reloading the skin on the GPU isn't straightforward at the moment. Unloading/updating an already loaded skin would require updating all uses of this skin in the client. This is currently only done by/when refreshing all skins, which would be inefficient to do when skins are downloaded while playing. Overall I want better skin resource handling as well, i.e. lazy loading and unloading of skins, but that's outside the scope of this PR.

Robyt3 commented on src/engine/shared/http.cpp:
```diff
@@ -126,6 +126,11 @@ bool CHttpRequest::ConfigureHandle(void *pHandle)
 	}
+	if(m_IfModifiedSince >= 0)
+	{
+		curl_easy_setopt(pH, CURLOPT_TIMEVALUE_LARGE, (curl_off_t)m_IfModifiedSince);
```
`CURLOPT_TIMEVALUE_LARGE` was added in curl 7.59.0, so if need to support earlier versions we need to use `CURLOPT_TIMEVALUE` also (which breaks in the year 2038).

heinrich5991 replied:
We should be good, even Debian 10 ships curl 7.64.0.

heinrich5991 commented on src/game/client/components/skins.cpp:
```diff
@@ -463,15 +425,191 @@ void CSkins::RandomizeSkin(int Dummy)
+		m_pSkins->Graphics()->LoadPng(m_ImageInfo, aPathReal, IStorage::TYPE_SAVE);
+	}
+
+	pGet->Wait();
```
This function unfortunately spins… #7811.

Robyt3 replied:
Fixed by #9145 and #9146.

heinrich5991 reviewed (approved):
Thanks!
